### PR TITLE
Prevent egui::Window from becoming larger than Viewport

### LINKED ISSUE #3987: Prevent egui::Window from becoming larger than Viewport
There is a need to prevent `egui::Window` from becoming larger than the Viewport.

### PR DESCRIPTION
**Pull request by rustbasic**
* Closes #3987  

There is a need to prevent egui::Window from becoming larger than the Viewport.

( This is basic. This alone is not enough. )

### CODE REVIEW

**Comment by YgorSouza** [comment]
@rustbasic if you put "Closes #3987" in the PR description like in the template, GitHub will automatically link the issue to the PR. That makes it easier to see what is being worked on.

**Comment by rustbasic** [comment]
Oh, thank you.

Also, if there are any problems with the source code I wrote, please let me know at any time.

File: crates/egui/src/containers/window.rs [hunk 1]
```diff
@@ -426,6 +426,12 @@ impl<'open> Window<'open> {
             0.0
         };
 
+        let viewport_rect = ctx.input(|i| i.viewport().inner_rect.unwrap_or(Rect::EVERYTHING));
```
Comment (emilk) [inline crates/egui/src/containers/window.rs:429]: I think it makes more sense to constrain it to `area.constrain_rect`

**Review by emilk — COMMENTED** [review]
Should we apply this to all windows or just resizable ones?
Maybe just those with scroll area?
Should it be opt-in or opt-out?
What is the effect of taking e.g. the `WidgetGallery` window in the demo, and putting it in a too small window?

File: crates/egui/src/containers/window.rs (hunk 1 above)
Comment (rustbasic) [reply crates/egui/src/containers/window.rs]: you're right. May be, It's what I wanted.

Comment (rustbasic) [reply crates/egui/src/containers/window.rs:430]: It's what I wanted right.

```
        let viewport_rect = ctx.input(|i| i.viewport().inner_rect.unwrap_or(Rect::EVERYTHING));
        let max_width = viewport_rect.width();
        let max_height = viewport_rect.height() - title_bar_height;
```

using `area.constrain_rect` does not limit window from continuing to grow.

**Comment by rustbasic** [comment]
> issue #3987
> 
> There is a need to prevent egui::Window from becoming larger than the Viewport.

File: crates/egui/src/containers/window.rs [hunk 2]
```diff
@@ -429,6 +429,12 @@ impl<'open> Window<'open> {
             (0.0, 0.0)
         };
 
+        let viewport_rect = ctx.input(|i| i.viewport().inner_rect.unwrap_or(Rect::EVERYTHING));
+        let max_width = viewport_rect.width();
+        let max_height = viewport_rect.height() - title_bar_height;
+        resize.max_size.x = resize.max_size.x.min(max_width);
+        resize.max_size.y = resize.max_size.y.min(max_height);
```
Comment (emilk) [inline crates/egui/src/containers/window.rs:436]: what I meant was this:

```suggestion
        let max_rect = area.constrain_rect.unwrap_or_else(|| ui.ctx().screen_rect);
        let max_width = max_rect.width();
        let max_height = max_rect.height() - title_bar_height;
        resize.max_size.x = resize.max_size.x.min(max_width);
        resize.max_size.y = resize.max_size.y.min(max_height);
```

**Pull request closed without merging**